src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,678 @@
+    bool operator== ( const node_address &other ) const {
+        return x == other.x && y == other.y && facing_dir == other.facing_dir;
+    }
+};
+
+struct node_address_hasher {
+    std::size_t operator()( const node_address &addr ) const {
+        std::int64_t val = addr.x;
+        val = ( val << 16 ) + addr.y;
+        val = ( val << 16 ) + addr.facing_dir;
+        return std::hash<int64_t> {}( val );
+    }
+};
+
+/*
+ * A node address annotated with its heuristic score, an approximation of how
+ * much it would cost to reach the goal from this node.
+ */
+struct scored_address {
+    struct node_address addr;
+    short score;
+    bool operator> ( const scored_address &other ) const {
+        return score >= other.score;
+    }
+};
+
+/*
+ * Data structure representing a navigation node that is known to be reachable. Contains
+ * information about the path to get there and enough information about the state of
+ * the vehicle and driver to predict which nodes can be reached on the next turn.
+ */
+struct navigation_node {
+    node_address prev;
+    short cost;
+    short tileray_steps;
+};
+
+class vehicle::autodrive_controller
+{
+    public:
+        explicit autodrive_controller( const vehicle &driven_veh, const Character &driver );
+        const Character &get_driver() {
+            return driver;
+        }
+        const auto_navigation_data &get_data() {
+            return data;
+        }
+        cata::optional<navigation_step> compute_next_step();
+
+    private:
+        const vehicle &driven_veh;
+        const Character &driver;
+        auto_navigation_data data;
+
+        bool check_obstacle( tripoint pt ) const;
+        void compute_obstacles();
+        vehicle_profile compute_profile( orientation facing ) const;
+        void compute_valid_positions( orientation facing );
+        void precompute_data();
+        scored_address compute_node_score( const node_address &addr, const navigation_node &node ) const;
+        bool compute_path();
+};
+
+
+static orientation normalize_orientation( int angle )
+{
+    return angle % NUM_ORIENTATIONS + ( angle < 0 ? NUM_ORIENTATIONS : 0 );
+}
+
+static int orientation_diff( int dir1, int dir2 )
+{
+    return normalize_orientation( dir1 - dir2 + NUM_ORIENTATIONS / 2 ) - NUM_ORIENTATIONS / 2;
+}
+
+static tripoint_abs_ms abs_vehicle_pos( const vehicle &veh )
+{
+    return tripoint_abs_ms( get_map().getabs( veh.global_pos3() ) );
+}
+
+static orientation approx_orientation( int dx, int dy )
+{
+    orientation ret = 0;
+    if( dy < 0 ) {
+        dx *= -1;
+        dy *= -1;
+        ret += NUM_ORIENTATIONS / 2;
+    }
+    if( dx < 0 ) {
+        std::swap( dx, dy );
+        dy *= -1;
+        ret += NUM_ORIENTATIONS / 4;
+    }
+    if( dy > dx ) {
+        ret += NUM_ORIENTATIONS / 4;
+        ret -= approx_orientation( dy, dx );
+    } else if( dy > 0 ) {
+        constexpr int table_size = 101;
+        static const std::array<int, table_size> atan_table = [] {
```

Comment:
Not atan2? I'm curious why not.

Reply:
I'm sorry, I don't quite understand what you mean. I'll take a shot at explaining.

The way this function works is it computes the tangent of the angle that the line segment {{0,0},{dx,dy}} makes with the x axis by simple division, normalizing the value to the range [0, 100] and performing an arctan operation on that value via a precomputed lookup table.
It's very fast because it just uses integer operations, no floating point.

I could certainly rewrite the whole thing using atan2, but that would be more expensive and this function is called from the inner loop of A* so I want it to be as fast as possible.

Reply:
I was curious as to the logic of not using atan2, which I'm accustomed to using due to better error-handling... but this isn't really a continuous distribution situation!